### ISSUE TITLE
Deleting old git branches

### ISSUE THREAD

**Issue by jkingdon** [metamath/set.mm#2476]
As people may have probably noticed there are a lot of old branches in https://github.com/metamath/set.mm . This is clutter to look at although I'm a little unclear about whether it has concrete effects like slowing down an initial clone of the repository.

What I've seen on other projects is some kind of bot to nag whoever had created branches to delete them if they are no longer needed. This is maybe is a bit more manual than would be ideal? What would people think about turning on the "Automatically delete head branches" setting? (this should be on the https://github.com/metamath/set.mm/settings page although I took this screenshot from another repository because I think only a few people have access to see it for this repository).

![image](https://user-images.githubusercontent.com/16878/152695701-6d146638-b43b-42ed-bd4b-74867c76ffce.png)

The documentation for this setting is here: https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/managing-the-automatic-deletion-of-branches . Although I didn't find a concise official statement of "this only affects branches within metamath/set.mm, not forks", that seems to be a true statement based on discussion at https://discuss.python.org/t/can-i-enable-the-automatically-delete-head-branch-feature-on-github/2088/6 and https://stackoverflow.com/questions/14485936/automatically-delete-git-branch-after-merge-to-master .

It should be relatively safe in the sense that even if a branch is deleted this way there is an "undelete this branch" button.

**Comment by digama0** [comment]
LGTM. I've enabled this option; we can revisit the issue if there are unforeseen side effects.

**Comment by benjub** [comment]
I'm all in favor. @digama0 , @jkingdon : will it delete them "from now on" or does it also take care of the "back log" ? In the former case, there will be some manual work to do...

**Comment by jkingdon** [comment]
> I'm all in favor. @digama0 , @jkingdon : will it delete them "from now on" or does it also take care of the "back log" ? In the former case, there will be some manual work to do...

This setting does nothing about the backlog.

**Comment by benjub** [comment]
> This setting does nothing about the backlog.

How can we deal with it ? On the page https://github.com/metamath/set.mm/branches, I can see the "delete" icon next to each one but I do not dare clicking it (plus, I may not have correct permission to do it).

**Comment by benjub** [comment]
I tried to delete one and it worked ("link-o2p2e4" by @david-a-wheeler).  Do you think, @jkingdon, that we can delete, say, all stale branches which have been merged (hence which are zero commits ahead of "develop") ?

**Comment by jkingdon** [comment]
> I tried to delete one and it worked ("link-o2p2e4" by @david-a-wheeler).

Yeah, I would expect anyone with write access would be able to do this. Unlike your own fork of the repository, this is a shared space.

> Do you think, @jkingdon, that we can delete, say, all stale branches which have been merged (hence which are zero commits ahead of "develop") ?

What I've usually seen is some level of "is this OK?" Like an email to the mailing list saying "I'll delete the following list on such-and-such a date unless you tell me you want me to keep one of them".

But especially for branches which have been merged, I'd be a little surprised if there is an objection.

**Comment by benjub** [comment]
Thanks. Done in https://groups.google.com/g/metamath/c/nplf76VwJ_o with target date starting on 19th February.

I'll let you see what to do for the other two metamath repositories (where I do not have write access nor am I asking for it): https://github.com/metamath/metamath-book and https://github.com/metamath/metamath-exe.

**Comment by jkingdon** [comment]
> Done in https://groups.google.com/g/metamath/c/nplf76VwJ_o

Oh cool. I didn't know there was a GitHub URL with the stale ones specifically.

**Comment by benjub** [comment]
@jkingdon : so, I deleted the branches which were: (1) stale, (2) merged, (3) zero commits ahead.  There are other stale branches which satisfy either (2) or (3). Can we delete them ? Actually, if you can check and delete them, do so. (I do not want to delete some by mistake.) See https://github.com/metamath/set.mm/branches/stale

**Comment by jkingdon** [comment]
> @jkingdon : so, I deleted the branches which were: (1) stale, (2) merged, (3) zero commits ahead. There are other stale branches which satisfy either (2) or (3). Can we delete them ? Actually, if you can check and delete them, do so. (I do not want to delete some by mistake.) See https://github.com/metamath/set.mm/branches/stale

Perhaps we are reaching the point where we don't really need to worry about it (at least not for branches created by other people)?

Take the example of https://github.com/metamath/set.mm/pull/707 . David made a pull request, Norm pushed back, and the resolution was "I'm closing for now and will investigate separately later." Will we want to refer to that again? I don't really know.

A case like https://github.com/metamath/set.mm/pull/95 which is merged but ahead is even harder to tease apart. I guess something was pushed to that branch after it was merged? Probably nothing important, but who can say other than the original author?

**Comment by david-a-wheeler** [comment]
I'd hate to delete #707 right away. I think it's worth breaking up & trying to incorporate a few of them. I just haven't gotten around to it. So I think for now, "we don't need to worry about it" is the easy course.

**Comment by benjub** [comment]
Then do not delete #707! There are probably some others you may deem deletable at once.  I agree that we are now at 39 branches, which fit on two pages, so it is more manageable than the few hundreds we had previously.

**Comment by digama0** [comment]
Regarding #95, the state of the branch is the same as commit cc08e95 (the squash commit for the PR on master): `git diff origin/sum1over2..cc08e95` returns nothing. So I think deleting the branch is fine. (This also applies to #1224, which is the only other case I could find of this situation.)

**Comment by benjub** [comment]
Both are due to @david-a-wheeler. If no objection, I'll delete them in a few days. There are also a few branches which are "0 commits ahead", but which are neither merged nor closed (there is a clickable "new pull request" button next to them). Can we delete them ?

**Comment by tirix** [comment]
My only stale branch is `proba`. It is 4 commits ahead, and especially two big commits [c9d5282](https://github.com/metamath/set.mm/commit/c9d52829d99c5711cf721749647baf61ad066f86) and [8a9d2c7](https://github.com/metamath/set.mm/commit/8a9d2c7bbde96c640dd770d6803db48a411c220c) were never merged.
I have to look at that.

**Comment by benjub** [comment]
> Regarding #95, the state of the branch is the same as commit [cc08e95](https://github.com/metamath/set.mm/commit/cc08e95ee2b01f4314001d3af9b690614dc5c2b3) (the squash commit for the PR on master): `git diff origin/sum1over2..cc08e95` returns nothing. So I think deleting the branch is fine. (This also applies to #1224, which is the only other case I could find of this situation.)

Just done (#95 and #1224).

**Comment by benjub** [comment]
> My only stale branch is `proba`. It is 4 commits ahead, and especially two big commits [c9d5282](https://github.com/metamath/set.mm/commit/c9d52829d99c5711cf721749647baf61ad066f86) and [8a9d2c7](https://github.com/metamath/set.mm/commit/8a9d2c7bbde96c640dd770d6803db48a411c220c) were never merged. I have to look at that.

@tirix : Have you ?

**Comment by benjub** [comment]
Among the stale branches (https://github.com/metamath/set.mm/branches/stale),
* some are zero commits ahead,
* some are the object of PRs which have been closed.

Can we delete those corresponding to either, the first, the second, or both items above ?

Also, there are two stale branches which are zero commits ahead and have been merged (#2337 and #2341 by @sctfn and @digama0). Probably they can be deleted ?

**Comment by sctfn** [comment]
Yeah, you can go and delete mine. On Aug 10, 2024, at 5:34 AM, Benoit ***@***.***> wrote:﻿
Among the stale branches (https://github.com/metamath/set.mm/branches/stale),

some are zero commits ahead,
some are the object of PRs which have been closed.

Can we delete those corresponding to either, the first, the second, or both items above ?
Also, there are two stale branches which are zero commits ahead and have been merged (#2337 and #2341 by @sctfn and @digama0). Probably they can be deleted ?

**Comment by tirix** [comment]
> > My only stale branch is `proba`. It is 4 commits ahead, and especially two big commits [c9d5282](https://github.com/metamath/set.mm/commit/c9d52829d99c5711cf721749647baf61ad066f86) and [8a9d2c7](https://github.com/metamath/set.mm/commit/8a9d2c7bbde96c640dd770d6803db48a411c220c) were never merged. I have to look at that.
> 
> @tirix : Have you ?

Actually I have not. 
Those commits correspond to an attempt at proving Descartes's Rule of Signs ([MM100 #100](https://us.metamath.org/mm_100.html#100))
The commits are probably of limited interest outside of a possible proof of the Rule of Signs.

The branch can probably not be merged as-is because of the many theorem renames.
The best course of action would probably be to make a new pull request with those changes.

**Comment by benjub** [comment]
> The best course of action would probably be to make a new pull request with those changes.

It'd be nice if you can do that, either rebasing or applying a patch to an up-to-date version of set.mm.

**Comment by benjub** [comment]
There are 7 stale branches that are zero commits ahead of `develop` (see https://github.com/metamath/set.mm/branches/stale).  **I am going to delete** the 6 ones other than `master` soon if no one objects here.
